staging/src/k8s.io/client-go/util/workqueue/delaying_queue_test.go

```diff
@@ -69,6 +285,247 @@ func TestSimpleQueue(t *testing.T) {
+	// Set up the waiting item...
+	q.AddAfter(first, 1*time.Second)
+	q.AssertState(t, QueueState{
+		waiting: []interface{}{first},
+	})
+
+	// Add an immediate item specifically disabling the active-waiting de-duplication...
+	q.AddWithOptions(second, DelayingOptions{PermitActiveAndWaiting: true})
+	q.AssertState(t, QueueState{
+		active:  []interface{}{second},
+		waiting: []interface{}{first},
+	})
+}
+
+// TestDelayedAddDropsWhenActivelyQueued calls for delayed adds should drop the delayed add if the item is aleady actively queued.
+func TestDelayedAddDropsWhenActivelyQueued(t *testing.T) {
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	q := NewTestDelayingQueueWithCustomClock(t, fakeClock, "")
+
+	first := "foo"
+	second := "foo"
+
+	// add first entry into the active queue
+	q.Add(first)
+	q.AssertState(t, QueueState{
+		active: []interface{}{first},
+	})
+
+	// try to add the same item with a delay, should have dropped the waiting request.
+	q.AddAfter(second, 1*time.Second)
+	q.AssertState(t, QueueState{
+		active: []interface{}{first},
+	})
+
+	// try to add the same item with a delay, buy with PermitActiveAndWaiting ...
+	q.AddWithOptions(second, DelayingOptions{Duration: 1 * time.Second, PermitActiveAndWaiting: true})
+	q.AssertState(t, QueueState{
+		active:  []interface{}{first},
+		waiting: []interface{}{second},
+	})
+}
+
+// TestOptionsTakeLongerTakeExisting - these options give the client an ability to perform a conditional add but
+// only when the item isn't already queued.  This fuctionality also applies to immediate adds made by the AddWithOptions
+// method.
+func TestOptionsTakeLongerTakeExisting(t *testing.T) {
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	q := NewTestDelayingQueueWithCustomClock(t, fakeClock, "")
+
+	first := "foo"
+	second := "foo"
+	third := "bar"
+
+	// add an item into the waiting queue.
+	q.AddAfter(first, 2*time.Second)
+
+	// attempt to immediately add an item using TakeLonger option
+	q.AddWithOptions(second, DelayingOptions{WhenWaiting: TakeLonger})
+	q.AssertState(t, QueueState{
+		waiting: []interface{}{first},
+	})
+
+	// attempt to immediately add an item using TakeExisting option
+	q.AddWithOptions(second, DelayingOptions{WhenWaiting: TakeExisting})
+	q.AssertState(t, QueueState{
+		waiting: []interface{}{first},
+	})
+
+	// attempt to add an item not queued should queue successfully...
+	q.AddWithOptions(third, DelayingOptions{WhenWaiting: TakeExisting})
+	q.AssertState(t, QueueState{
+		active:  []interface{}{third},
+		waiting: []interface{}{first},
+	})
+
+	// step past the original delayed item and process it from the active queue
+	fakeClock.Step(5 * time.Second)
+	if err := waitForAdded(q, 2); err != nil {
+		t.Fatalf("the first item should have been queued")
+	}
+	q.AssertState(t, QueueState{
+		active: []interface{}{first, third},
+	})
+	// pop them both off the queue
+	item, _ := q.Get()
+	q.Done(item)
+	item, _ = q.Get()
+	q.Done(item)
```

Comment:
would not hurt to compare these items here as well

Reply:
Its not needed because we have just run an AssertState which has compared the items - this is only to unqueue them.